### PR TITLE
("what-to-build") Modularize different run platforms (i.e. bitstream builds)

### PR DESCRIPTION
**Pull request by abejgonzalez**
*Goals:*
- *what things are build* part of #848
- Abstract out AWS F1 specific parts of the RTL, driver, and bitstream creation

*Misc Notes*
Not that this is not heavily tested at all

*Extended Changes*

This removes `buildafi.py` and moves its functionality (and some more) into `bitbuilder.py`. Bitbuilder classes manage how a specific run platform should build the RTL, driver, and bitstream. Bitbuilders are assigned to a build using `fpgaplatform=<BIT BUILDER CLASS>` in the `build_recipes.ini`. 

TODO:
- [ ] ~~Need to change HWDB schema so that it can be used with different platforms. Maybe rename/reuse the bitbuilder so that it also is attached to a HWDB entry (and it will also be used when ready to run things)?~~ Maybe this is being done as part of "what-to-run"?
- [x] Rename `buildafi` to `buildbitstream` in the manager and add deprecation warnings to `buildafi`.
- [ ] ~~Determine if there is a cleaner way to specify a build recipe~~

#### Related PRs / Issues

Taken and expanded on from #848 

#### UI / API Impact

See description

#### Verilog / AGFI Compatibility

N/A

### Contributor Checklist
- [ ] Did you set dev as the base branch?
- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous prints/debugging code?
- [ ] Did you state the UI / API impact?
- [ ] Did you specify the Verilog / AGFI compatibility impact?

- [ ] (If applicable) Did you regenerate and publicly share default AGFIs?

- [ ] (If applicable) Did you mark the PR as "Please Backport"?
- [ ] (On merge) Did you update release notes in the dev-to-master PR ?

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

**Comment by abejgonzalez** [comment]
While not a part of this PR, we should probably look into the additive functionality given by launchrunfarm - https://github.com/firesim/firesim/pull/716#discussion_r801916779

**Comment by abejgonzalez** [comment]
Looks like XRT now supports static linked libraries https://github.com/Xilinx/XRT/pull/6217. If possible this should get tested here, otherwise we can try this out later.

**Comment by abejgonzalez** [comment]
(Not implemented) I envision that the build recipe will change to something like this:

```
<NAME>:
    ...
    fpga-platform:
        type: <HUMAN READABLE NAME CORRESPONDING TO THE BITBUILDER CLASS>
        args:
            <K/B PAIRS OF ARGS>
```

So for the default F1 case I can see something like:

```
<NAME>:
    ...
    fpga-platform:
        type: aws-ec2-f1
        args:
            s3-bucket-name: ...
            post-build-hook: ...
```

If we want to make things a bit cleaner as well we can move the `fpga-platform` into its own file (like the build hosts) so that the platform can be shared across builds.

**Comment by timsnyder-siv** [comment]
@abejgonzalez is this ready to be reviewed or should I wait until #850 is merged?

File: deploy/firesim
Comment (timsnyder-siv) [inline deploy/firesim:271]: Given that managerinit warns the user to run awsinit, I think we should just bite the bullet, update the docs and not introduce a new option `-o` for something that is being deprecated as the option is being added.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:152]: In the docstring for this method, change managerinit to awsinit.

File: deploy/buildtools/buildconfig.py
Comment (timsnyder-siv) [inline deploy/buildtools/buildconfig.py:115]: Why do you use the raw python class names for BitBuilder classes but for BuildHostDispatcher, you use a different string name and build up a mapping?  I would expect that you would want to use the same pattern in both places.

Comment (timsnyder-siv) [reply deploy/buildtools/buildconfig.py:115]: Also, I think that we want to be able to plugin other bitbuilders without having to pollute the firesim codebase.  We should move away from only being able to lookup BitBuilderDispatchers in the buildtools.bitbuilder module.

**Review by timsnyder-siv — COMMENTED** [review]
This PR is much easier to follow and doesn't want to crash my browser when I look at it.  Thanks.

**Review by davidbiancolin — DISMISSED** [review]
I'm not convinced we want to expose separate tasks (awsinit, <myhosttype>init) for each host-type we support. Instead maybe manager init should take a list of host-types for which specialized `<host>init` methods should called. These methods can be provided using the same approach we're going to use to supply implementations for builds and simulations. This is fine for now though.

**Comment by timsnyder-siv** [comment]
I'm starting to work on rebasing what-to-build to the tip of loca-fpga, rolling in the stuff from #850

File: deploy/buildtools/buildconfig.py
Comment (timsnyder-siv) [reply deploy/buildtools/buildconfig.py:78]: I didn't come across this while merging and getting things to work again.  We probably still want to make the construction of the `fpga_bit_builder_dispatcher_class_name` work the same way as the other places where we're using the inheritors map.  Since that code is moving to `deploy/utils/inheritors.py` in #1028, perhaps we should merge this as is and clean it up to be consistent with the new code location in #1028?

Comment (abejgonzalez) [reply deploy/buildtools/buildconfig.py:78]: That PR will take longer to merge in so I think this plan is good (merge this 1st - cleanup in the other PR)

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim]: I started this with the addition of the `--platform` flag and moving `awsinit()` into `awstools.py`

File: deploy/buildtools/bitbuilder.py
Comment (timsnyder-siv) [reply deploy/buildtools/bitbuilder.py:53]: I didn't do this as a part of the initial merge.  Let me know if you would still like it done before we merge this PR.

Comment (abejgonzalez) [reply deploy/buildtools/bitbuilder.py:53]: If we rename this, I think we should recombine the `replace_rtl` and `build_driver` steps into one. I originally split things up to be a bit clearer in the code but Im fine with either approach.

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [inline deploy/buildtools/bitbuilder.py:75]:
```suggestion
        rootLogger.info(f"Building Verilog for {self.build_config.get_chisel_triplet()}")
```

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [inline deploy/buildtools/bitbuilder.py:89]:
```suggestion
        rootLogger.info(f"Building FPGA driver for {self.build_config.get_chisel_triplet()}")
```

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [inline deploy/buildtools/bitbuilder.py:26]: Can we avoid using Fabric to determine this path?

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [inline deploy/buildtools/bitbuilder.py:52]: Can we somehow assert that this has to run on localhost? If this is too complicated, I think we should mention what functions in this class are designed to run locally vs anywhere

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [inline deploy/buildtools/bitbuilder.py:138]:
```suggestion
                local_dir=local_awsfpga_dir,
                remote_dir=dest_f1_platform_dir,
                ssh_opts="-o StrictHostKeyChecking=no",
                exclude=["hdk/cl/developer_designs/cl_*"],
                extra_opts="-l", capture=True)
            rootLogger.debug(rsync_cap)
            rootLogger.debug(rsync_cap.stderr)
            rsync_cap = rsync_project(
                local_dir=f"{local_awsfpga_dir}/{fpga_build_postfix}/*",
                remote_dir=f'{dest_awsfpga_dir}/{fpga_build_postfix}',
                exclude=["build/checkpoints"],
                ssh_opts="-o StrictHostKeyChecking=no",
                extra_opts="-l", capture=True)
```

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [inline deploy/buildtools/bitbuilder.py:293]: I think in this PR we should somehow create a section that changes based on the platform it builds for. I.e change the HWDB to be something like:

```
<NAME>:
  deploy_triplet_override: null
  custom_runtime_config: null
  platform_specific:
      type: aws-fpga
      args:
           agfi: ....
```

Comment (abejgonzalez) [reply deploy/buildtools/bitbuilder.py:293]: For now, the PR would just support the AWS version of whatever the API would be but this would make the Vitis integration much easier.

**Review by abejgonzalez — COMMENTED** [review]
Overall seems good to me. Was a `buildbitstream` run done?

File: deploy/awstools/awstools.py
Comment (abejgonzalez) [inline deploy/awstools/awstools.py:143]:
```suggestion
def awsinit() -> None:
```

File: deploy/awstools/awstools.py
Comment (abejgonzalez) [reply deploy/awstools/awstools.py:186]: +1

File: deploy/awstools/awstools.py
Comment (abejgonzalez) [inline deploy/awstools/awstools.py:196]: This should probably just error like in #1037 when the error code isn't 0, 28

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [reply deploy/buildtools/bitbuilder.py:26]: This can be ignored for now.

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [inline deploy/buildtools/bitbuilder.py:53]:
```suggestion
        """Generate Verilog from build config. Should run on the manager host."""
```

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [reply deploy/buildtools/bitbuilder.py:53]: Sounds like we can do this later.

File: deploy/buildtools/bitbuilder.py
Comment (abejgonzalez) [reply deploy/buildtools/bitbuilder.py:293]: Im fine with punting this to a later PR (since we also need to add in the URI changes)

File: deploy/firesim
Comment (abejgonzalez) [inline deploy/firesim:103]: Note for later, we should unify how we add "ERROR", "WARNING" to strings

**Review by abejgonzalez — COMMENTED** [review]
Added small changes but this PR is good to go for me. I'll test this against the Vitis stuff once this goes in.

**Comment by abejgonzalez** [comment]
Before this PR gets merged can we get CI to run on it.

File: deploy/buildtools/buildconfigfile.py
Comment (timsnyder-siv) [reply deploy/buildtools/buildconfigfile.py:167]: I didn't think I deleted it.  Thanks for pointing it out.  Should be added back.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:103]: I agree.  And it probably should be done with the logging formatter(s) instead of at each caller.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:149]: The only reason why it is there is so that it get's registered as a legal taskname for argparse to accept.  Otherwise, argparse will error if you give it and that's the same as not having the deprecation.  I could add to the comment to explain that.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:204]: I use it fairly often when I push hard on Vivado.  I agree that thought needs to be put into if/how to generalize this.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:336]: lol.  NO.  I want to call it `bitream`.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:192]:
```suggestion
            config_file: BuildConfigFile
```

File: deploy/buildtools/bitbuilder.py
Comment (timsnyder-siv) [reply deploy/buildtools/bitbuilder.py:26]: I think it would be better to do something with `__file__`like `return Path(__file__).parent.parent` and perhaps assert that what you looked up is a directory named `deploy` but I agree that it can wait.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [reply deploy/awstools/awstools.py:196]: Yes.  I figured #1037 supercedes this and you'll take care of it.

File: deploy/buildtools/buildconfig.py
Comment (abejgonzalez) [inline deploy/buildtools/buildconfig.py:74]: None of the recipes changed to add this. I would either 1. add it to each recipe or 2. for now just set it to "F1BitBuilder" by default. We can modify this later.

File: deploy/firesim
Comment (abejgonzalez) [reply deploy/firesim:104]: https://github.com/firesim/firesim/issues/1047

File: deploy/sample-backup-configs/sample_config_hwdb.yaml
Comment (timsnyder-siv) [inline deploy/sample-backup-configs/sample_config_hwdb.yaml:33]: Does this mean that you regenerated AGFI's with this PR or was the merge somehow messed up?  Just curious.

File: deploy/ssh-setup.sh
Comment (timsnyder-siv) [inline deploy/ssh-setup.sh:36]: I think that regardless of whether you are on F1 or a local cluster or some other cloud provider, you are going to be using Fabric to SSH to other hosts and you will want key-based auth to work.  I don't think we should make this message F1 specific but we could clean this up in a follow-on.

File: deploy/sample-backup-configs/sample_config_hwdb.yaml
Comment (abejgonzalez) [reply deploy/sample-backup-configs/sample_config_hwdb.yaml:33]: I regenerated AGFIs with the most recent commits just to verify that everything was working. I then noticed that there was a duplicate HWDB entry so I figured I would just update everything.

**Review by sagark — APPROVED** [review]
LGTM

**Comment by abejgonzalez** [comment]
Rerunning CI since 1. I forgot to enable FPGA sim and 2. Looks like the UI was hanging (not a fault of this PR)

**Pull request merged**
